### PR TITLE
Add new API URLs

### LINKED ISSUE #45: IPV6 or IPV4
Currently, in the browser version if v6 function is called and the API request returns a v4 version, then an error is made and we can't get the IP.

It'd be good if there's a single function to return IPV6/IPV4, if v6 is available :

```
console.log(await publicIp.get());
//=> '46.5.21.123' or v6 if available (preferred)
```

### PR DESCRIPTION
**Pull request by subins2000**
Fixes #45 
Added new general HTTPs APIs from which IP is retrieved with a regex pattern. Adds a JSON object to do that :
```
general: {
  'https://www.cloudflare.com/cdn-cgi/trace': 'ip(=)(.*?)\n',
  'https://ip-api.io/api/json': 'ip"(.*?)"(.*?)"'
}
```

### CODE REVIEW

**Comment by sindresorhus** [comment]
This will need some tests.

**Comment by subins2000** [comment]
The current browser tests are pretty basic. Shall I rewrite it with [airtap](https://github.com/airtap/airtap) ?

**Comment by sindresorhus** [comment]
> The current browser tests are pretty basic. Shall I rewrite it with airtap ?

Sure

File: browser.js [hunk 1]
```diff
@@ -59,6 +64,27 @@ queryHttps.cancel = () => {
 	xhr.abort();
 };
 
+const v6or4 = async options => {
+	let ip;
```
Comment (sindresorhus) [inline browser.js:68]: This doesn't need to be defined at this scope.

File: browser.js [hunk 2]
```diff
@@ -13,7 +13,12 @@ const urls = {
 	v6: [
 		'https://ipv6.icanhazip.com/',
 		'https://api6.ipify.org/'
-	]
+	],
+	// URL => regex which match IP in response
+	v6or4: {
+		'https://www.cloudflare.com/cdn-cgi/trace': 'ip(=)(.*?)\n',
+		'https://ip-api.io/api/json': 'ip"(.*?)"(.*?)"'
+	}
```
[the remark this answers is not in the document]
Comment (subins2000) [reply browser.js:21]: Then I'll have to modify those from array to objects like above with the regex pattern.

File: browser.js (hunk 1 above)
Comment (subins2000) [reply browser.js:68]: Sorry, but I don't understand what you meant

Comment (sindresorhus) [reply browser.js:68]: It's not used outside the try catch, so it can be moved to line 75.

**Comment by subins2000** [comment]
I've added tests for browser. It works good to me (my ISP doesn't give me a IPv6 address). Can you please test it if you have IPv6 ?

As for Node, I'm thinking of doing it in a try catch block (contrasting from the way it's done in browser), if ipv6 methods fail, use v4. What do you think ?

**Comment by sindresorhus** [comment]
> I've added tests for browser. It works good to me (my ISP doesn't give me a IPv6 address). Can you please test it if you have IPv6 ?

I don't have IPv6 where I'm at right now.

> I see the linter has 'prefer-await-to-then', but the code is shorter in test-browser.js with .then() and more cleaner. May I suggest to remove this lint ?

No, the lint rule is there for a reason. It generally leads to more readable code.

> As for Node, I'm thinking of doing it in a try catch block (contrasting from the way it's done in browser), if ipv6 methods fail, use v4. What do you think ?

:+1:

File: browser.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply browser.js:21]: 👍

**Comment by sindresorhus** [comment]
@subins2000 Bump

**Comment by subins2000** [comment]
Hey @sindresorhus been busy, but I'll try to work on this

**Comment by subins2000** [comment]
Hi @sindresorhus 

I've realized that instead of using a v6or4 function, it's better to explicitly retrieve the format of IP address needed. The new commits add the CloudFlare API & general IP retrieval APIs as a fallback if the first ones fail (browser). 

In case of dual IPv6 & IPv4 stack network interfaces, these new APIs would return one or the another (browser), but if the explicit first 2 fails, it's good to have a backup (even if they might return the opposite IP address).

File: test-browser.js [hunk 3]
```diff
@@ -2,12 +2,33 @@
 // $ browserify test-browser.js | pbcopy
 'use strict';
 const publicIp = require('./browser');
+const test = require('tape');
 
-(async () => {
-	console.log('IP:', await publicIp.v4());
-	console.log('IP:', await publicIp.v4({
+let v6Address = null;
```
Comment (sindresorhus) [inline test-browser.js:7]:
```suggestion
let v6Address;
```

Comment (sindresorhus) [reply test-browser.js:7]: And why are you setting it at the top-level? It's not used anywhere else.

File: test-browser.js [hunk 4]
```diff
@@ -2,12 +2,33 @@
...
+test('v6', async t => {
+	try {
+		const ip1 = await publicIp.v6();
+		v6Address = await publicIp.v6();
+
+		t.equal(ip1, v6Address);
+		t.end();
+	} catch (error) {
+		if (error === 'Couldn\'t find your IP') {
+			v6Address = null;
```
Comment (sindresorhus) [inline test-browser.js:18]: I'm not sure what the point of setting this to null is.

Shouldn't this case fail the test?

File: index.js [hunk 5]
```diff
@@ -170,6 +180,28 @@ const queryHttps = (version, options) => {
 				}
 			}
 
+			// Try regex URLs
+			const urls_ = regexUrls;
+			for (const url of Object.keys(urls_)) {
```
Comment (sindresorhus) [inline index.js:185]: Use `Object.entries`

File: index.js [hunk 6]
```diff
@@ -170,6 +180,28 @@ const queryHttps = (version, options) => {
...
+				try {
+					const gotPromise = got(url, requestOptions);
+					cancel = gotPromise.cancel;
+
+					// eslint-disable-next-line no-await-in-loop
+					const response = await gotPromise;
```
Comment (sindresorhus) [inline index.js:191]: There's a lot of duplication here.

File: index.js [hunk 7]
```diff
@@ -57,6 +57,16 @@ const dnsServers = [
 	}
 ];
 
+// URL => regex which match IP in response
```
Comment (sindresorhus) [inline index.js:60]:
```suggestion
// URL and regex which matches the IP in the response.
```

File: index.js [hunk 8]
```diff
@@ -170,6 +180,28 @@ const queryHttps = (version, options) => {
...
+			const urls_ = regexUrls;
+			for (const url of Object.keys(urls_)) {
+				try {
+					const gotPromise = got(url, requestOptions);
+					cancel = gotPromise.cancel;
+
+					// eslint-disable-next-line no-await-in-loop
+					const response = await gotPromise;
+
+					const ip = (response.body || '').trim().match(urls_[url]);
```
Comment (sindresorhus) [inline index.js:193]: I prefer using `.exec` over `.match`.

File: browser.js [hunk 9]
```diff
@@ -13,7 +13,14 @@ const urls = {
 	v6: [
 		'https://ipv6.icanhazip.com/',
 		'https://api6.ipify.org/'
-	]
+	],
+	// These URLs accept both IPv6/IPv4 connectivity
+	// Response might change for dual IPv6/IPv4 stack network interface
+	// URL => regex which match IP in response
+	general: {
+		'https://www.cloudflare.com/cdn-cgi/trace': 'ip(=)(.*?)\n',
```
Comment (sindresorhus) [inline browser.js:21]:
```suggestion
		'https://www.cloudflare.com/cdn-cgi/trace': 'ip(=)(.*?)',
```
The body is trimmed anyway.

Comment (sindresorhus) [reply browser.js:21]: We should not duplicate all this information. Move the shared info into a file that is imported by both the browser and Node.js implementation.

Comment (sindresorhus) [reply browser.js:21]: Use named regex groups.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: test-browser.js (hunk 4 above)
Comment (subins2000) [reply test-browser.js]: Sorry, my bad, it was the leftover code from the previous `v6or4` function test change. I should have taken a second look

File: api-urls-regex.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// URL and regex which matches the IP in the response.
```
Comment (sindresorhus) [inline api-urls-regex.js:1]: I would call the file `shared.js` and use a named export instead.

This also needs to be added to `files` in package.json.

File: index.js [hunk 11]
```diff
@@ -160,40 +152,31 @@ const queryHttps = (version, options) => {
 
 			const urls = [].concat.apply(type[version].httpsUrls, options.fallbackUrls || []);
 
-			for (const url of urls) {
+			// Make object with url as key and pattern as value
+			const urls_ = {
+				...urls.reduce((result, item) => {
```
Comment (sindresorhus) [inline index.js:157]: Don't use `reduce`. It makes code harder to read. A `for-of` loop is more readable.

File: index.js [hunk 12]
```diff
@@ -160,40 +152,31 @@ const queryHttps = (version, options) => {
...
+					result[item] = null;
```
Comment (sindresorhus) [inline index.js:158]: Don't use `null`.

https://github.com/sindresorhus/meta/discussions/7

**Comment by sindresorhus** [comment]
Are there any breaking changes in this PR? Do we need any docs updates?

**Comment by sindresorhus** [comment]
@subins2000 Bump

**Comment by sindresorhus** [comment]
@subins2000 Final bump ;)

**Comment by subins2000** [comment]
@sindresorhus Hi Sindre, unfortunately I don't have time to work on this PR anymore. Hopefully someone else can pick up the pieces in future. You can close this PR at your discretion.

Thanks

**Pull request closed without merging**